SDL/sound.c

```diff
@@ -0,0 +1,319 @@
+  emin = BUFFER_SAMPLES;
+  emax = -BUFFER_SAMPLES;
+#endif
+  adjust += Sound_FudgeRate;
+  /* Limit to within range [0.5,2.0] */
+  if (adjust < 1<<23)
+  {
+    adjust = 1<<23;
+  }
+  else if (adjust > 2<<24)
+  {
+    adjust = 2<<24;
+  }
+  Sound_FudgeRate = adjust;
+}
+
+void Sound_HostBuffered(SoundData *buffer,int32_t numSamples)
+{
+  numSamples <<= 1;
+  Sound_Lock();
+  int32_t local_buffer_in = sound_buffer_in;
+  int32_t used = local_buffer_in-sound_buffer_out;
+  int32_t out = sound_buffer_hostout;
+  int32_t underflows = sound_underflows;
+  sound_underflows = 0;
+  LOG_EVENT("Sound_HostBuffered",numSamples);
+  Sound_Unlock();
+  int32_t ofs = local_buffer_in & sound_buff_mask;
+  int32_t buffree = BUFFER_SAMPLES-MAX(ofs,used);
+
+  local_buffer_in += numSamples;
+
+  if (underflows)
+  {
+    fprintf(stderr, "*** sound underflow x%d! ***\n", underflows);
+  }
+
+  adjust_fudgerate(used, out);
+
+  Sound_Lock();
+  sound_buffer_in = local_buffer_in;
+  Sound_Unlock();
+}
+
+static void Sound_Callback(void *userdata, Uint8 *stream, int len)
+{
+  static int32_t local_buffer_out = 0;
+  while (len) {
+    int32_t avail;
+
+    Sound_Lock();
+    sound_buffer_hostout += len/sizeof(SoundData);
+    sound_buffer_out = local_buffer_out;
+    LOG_EVENT("Sound_Callback",-(len/sizeof(SoundData)));
+    avail = sound_buffer_in-local_buffer_out;
+    if (avail * sizeof(SoundData) < len)
+      sound_underflows++;
+    Sound_Unlock();
+
+    if (avail) {
+      int32_t ofs = local_buffer_out & sound_buff_mask;
+
+      if(ofs + avail > BUFFER_SAMPLES) {
+        /* We're about to wrap */
+        avail = BUFFER_SAMPLES-ofs;
+      }
+
+      if (avail * sizeof(SoundData) > len) {
+        avail = len / sizeof(SoundData);
+      }
+
+      memcpy(stream, sound_buffer + ofs,
+            avail * sizeof(SoundData));
+
+      local_buffer_out += avail;
+
+      stream += avail * sizeof(SoundData);
+      len -= avail * sizeof(SoundData);
+    } else {
+      memset(stream, 0, len);
+      return;
+    }
+  }
+}
+
+int
+Sound_InitHost(ARMul_State *state)
+{
+  SDL_AudioSpec desired, obtained;
+
+  desired.freq = 44100;
+  desired.format = AUDIO_S16SYS;
+  desired.channels = 2;
+  desired.samples = 512;
+  desired.callback = Sound_Callback;
+  desired.userdata = NULL;
+
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+  device_id = SDL_OpenAudioDevice(NULL, 0, &desired, &obtained,
+                                  SDL_AUDIO_ALLOW_FREQUENCY_CHANGE | SDL_AUDIO_ALLOW_SAMPLES_CHANGE);
+  if (device_id == 0)
+#else
+  if (SDL_OpenAudio(&desired, &obtained) < 0)
+#endif
+  {
+    fprintf(stderr, "Could not open audio device: %s\n", SDL_GetError());
+    return -1;
+  }
+
+#ifdef SOUND_LOGGING
+  logfile = fopen("soundlog.csv","w");
+  if (logfile)
+  {
+    fprintf(logfile,"event,time,used,delta\n");
+  }
+  fprintf(stderr, "Timer frequency %llu\n", SDL_GetPerformanceFrequency());
+  logt0 = SDL_GetPerformanceCounter();
```

Comment:
I know this is a small thing since it's only used for debugging, but could you add a basic fallback for these two functions when building with SDL 1.2?

Reply:
Good catch - fixed now.